### PR TITLE
[DuckPlayer] - Base Implementation

### PR DESCRIPTION
**Pull request by afterxleep**
Task/Issue URL: https://app.asana.com/0/1204099484721401/1207593922684284/f
Tech Design URL: https://app.asana.com/0/1204099484721401/1207530480716552/f

**Description**:
- Implements base DuckPlayer functionality based on macOS code
- Abstract navigation logic and Adds a YoutubeNavigationHandler to TabViewController that is used when DuckPlayer is enabled
- Adds App Settings to Manage DuckPlayer State
- Adds Tests for the Navigation Handler to ensure redirect works as expected and completion calls are always called

**Steps to test this PR**:
1.  Make yourself internal User
2.  Go to Settings > DuckPlayer and Select 'Always'
3. Navigate to youtube.com and tap on a video
4. Confirm video loads in DuckPlayer
5. Go Back
6. Confirm you're taken back to the Main YouTube page
7. Enter a Duck URL in the Omnibar, like. duck://player/0HZPyhnErFo
8. Confirm video is loaded in DuckPlayer
9. Go back to settings and disable DuckPlayer
10. Confirm YouTube video links no longer load in Duck player

⚠️  Known issues / Work in progress

This is the first of several PRs and is still missing features

- AlwaysASK mode does not work
- Youtube Video overlays do not work
- DuckPlayer UI uses macOS HTML views (mobile views are WIP)
- DuckPlayer UI buttons do not work
- Settings Design is not fully implemented

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)

### CODE REVIEW

File: DuckDuckGo.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved [hunk 1]
```diff
@@ -138,10 +138,10 @@
     {
       "identity" : "swift-argument-parser",
       "kind" : "remoteSourceControl",
-      "location" : "https://github.com/apple/swift-argument-parser.git",
+      "location" : "https://github.com/apple/swift-argument-parser",
       "state" : {
-        "revision" : "0fbc8848e389af3bb55c182bc19ca9d5dc2f255b",
-        "version" : "1.4.0"
+        "revision" : "46989693916f56d1186bd59ac15124caef896560",
+        "version" : "1.3.1"
```
Comment (Bunn) [inline DuckDuckGo.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved:144]: is this intended?

File: DuckDuckGo/DuckPlayer/DuckPlayer.swift [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+import WebKit
+import UserScript
+
+enum DuckPlayerMode: Equatable, Codable, CustomStringConvertible, CaseIterable {
+    case enabled, alwaysAsk, disabled
+    
+    var description: String {
+        switch self {
+        case .enabled:
+            return "Always"
```
Comment (Bunn) [inline DuckDuckGo/DuckPlayer/DuckPlayer.swift]: Should this use static localizable strings?

File: DuckDuckGo/DuckPlayer/DuckPlayer.swift [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+        from origin: YoutubeOverlayUserScript.MessageOrigin
+    ) -> (_ params: Any, _ message: UserScriptMessage) -> Encodable? {
+
+        return { [weak self] params, _ -> Encodable? in
+            guard let self else {
+                return nil
+            }
+            guard let userValues: UserValues = DecodableHelper.decode(from: params) else {
+                assertionFailure("YoutubeOverlayUserScript: expected JSON representation of UserValues")
+                return nil
```
Comment (Bunn) [inline DuckDuckGo/DuckPlayer/DuckPlayer.swift:161]: Maybe add a pixel here?

File: DuckDuckGo/DuckPlayer/DuckPlayerURLExtension.swift [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+              timestamp.matches(regex)
+        else {
+            return self
+        }
+        return appendingParameter(name: "t", value: timestamp)
+    }
+}
+
+extension CharacterSet {
+    static let youtubeVideoIDNotAllowed = CharacterSet(charactersIn: "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_").inverted
```
Comment (Bunn) [inline DuckDuckGo/DuckPlayer/DuckPlayerURLExtension.swift:127]: Where this data set comes from?

File: DuckDuckGo/DuckPlayer/YouTubePlayerNavigationHandler.swift [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+    }
+    
+    func goBack(webView: WKWebView) {
+        guard let backURL = webView.backForwardList.backItem?.url,
+                backURL.isYoutubeVideo,
+                backURL.youtubeVideoParams?.videoID == webView.url?.youtubeVideoParams?.videoID else {
+            webView.goBack()
+            return
+        }
+        webView.goBack(skippingHistoryItems: 2)
```
Comment (Bunn) [inline DuckDuckGo/DuckPlayer/YouTubePlayerNavigationHandler.swift:117]: Why goBack 2? Maybe add a comment with the reason why?

**Review by Bunn — APPROVED** [review]
LGTM, added some comments but nothing looks critical

File: DuckDuckGo/DuckPlayer/YouTubePlayerNavigationHandler.swift (hunk 5 above)
Comment (afterxleep) [reply DuckDuckGo/DuckPlayer/YouTubePlayerNavigationHandler.swift:117]: We Go from:

- Youtube list (youtube.com)
- Youtube video link (youtube.com/watch?=) which Redirects to Duck Player (duck://player/xYz)

The standard goBack() would take you to the Youtube video link, and you'll end in a loop

File: DuckDuckGo/DuckPlayer/DuckPlayerURLExtension.swift (hunk 4 above)
Comment (afterxleep) [reply DuckDuckGo/DuckPlayer/DuckPlayerURLExtension.swift:127]: That's a very good question.  I have no idea. 

This extension came out directly from macOS implementation and we'll eventually move it to BSK.

File: DuckDuckGo/DuckPlayer/DuckPlayer.swift (hunk 3 above)
Comment (afterxleep) [reply DuckDuckGo/DuckPlayer/DuckPlayer.swift:161]: Good Call.   I added to the pixel list here: https://app.asana.com/0/1204099484721401/1207593922684294

File: DuckDuckGo/DuckPlayer/DuckPlayer.swift (hunk 2 above)
Comment (afterxleep) [reply DuckDuckGo/DuckPlayer/DuckPlayer.swift]: Done.

**Pull request merged**